From bb994215b8e415cd5815dacf6bea95012ded60fb Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Wed, 8 Mar 2023 10:09:10 +0530
Subject: [PATCH 1/2] hostapd: add support to use underlying hw_idx for multi-
 hardware under single phy mode

During multi-hardware under singly phy scenario, single channel list is
maintained for a given band which does not clearly tells how many different
underlying hardwares are being used to form that band. For example - in
5 GHz split hardware case, two underlying hardwares 5 GHz Low and 5 GHz High
are being used to enable full 5 GHz band. Now, when channel is selected from
list of channels, it is possible that a channel can be picked which the
hardware does not support (lower chanels can be picked for 5 GHz high
hardware) which is wrong.

However, the kernel does tells us about the underlying hardwares and the
list of supported frequencies.

Hence, add support to get the advertised hw_idx and the corresponding list
of supported frequencies and then decide what is the current hw_idx being
used.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 src/ap/ap_drv_ops.c               |   9 ++
 src/ap/ap_drv_ops.h               |   3 +
 src/ap/hostapd.c                  |   7 ++
 src/ap/hostapd.h                  |   3 +
 src/ap/hw_features.c              |  77 ++++++++++++++++
 src/ap/hw_features.h              |  14 +++
 src/drivers/driver.h              |  32 +++++++
 src/drivers/driver_nl80211.c      |   1 +
 src/drivers/driver_nl80211.h      |   2 +
 src/drivers/driver_nl80211_capa.c | 144 ++++++++++++++++++++++++++++++
 src/drivers/nl80211_copy.h        |  41 +++++++++
 tests/fuzzing/ap-mgmt/ap-mgmt.c   |   2 +
 12 files changed, 335 insertions(+)

--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -1094,3 +1094,12 @@ int hostapd_drv_set_secure_ranging_ctx(s
 	return hapd->driver->set_secure_ranging_ctx(hapd->drv_priv, &params);
 }
 #endif /* CONFIG_PASN */
+
+struct hostapd_multi_hw_info *
+hostapd_get_multi_hw_info(struct hostapd_data *hapd, u8 *num_multi_hws)
+{
+	if (hapd->driver == NULL ||
+	    hapd->driver->get_multi_hw_info == NULL)
+		return NULL;
+	return hapd->driver->get_multi_hw_info(hapd->drv_priv, num_multi_hws);
+}
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -472,4 +472,7 @@ hostapd_drv_register_frame(struct hostap
 }
 #endif /* CONFIG_TESTING_OPTIONS */
 
+struct hostapd_multi_hw_info *
+hostapd_get_multi_hw_info(struct hostapd_data *hapd, u8 *num_multi_hws);
+
 #endif /* AP_DRV_OPS */
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -607,6 +607,9 @@ void hostapd_cleanup_iface_partial(struc
 	ap_list_deinit(iface);
 	sta_track_deinit(iface);
 	airtime_policy_update_deinit(iface);
+	hostapd_free_multi_hw_info(iface->multi_hw_info, iface->num_multi_hws);
+	iface->multi_hw_info = NULL;
+	iface->current_hw_info = NULL;
 }
 
 
@@ -2333,6 +2336,10 @@ static int hostapd_setup_interface_compl
 			   hostapd_hw_mode_txt(iface->conf->hw_mode),
 			   iface->conf->channel, iface->freq);
 
+		if (hostapd_set_current_hw_info(iface, iface->freq)) {
+			wpa_printf(MSG_ERROR, "Failed to get operating hw mac id");
+			goto fail;
+		}
 #ifdef NEED_AP_MLME
 		/* Handle DFS only if it is not offloaded to the driver */
 		if (!(iface->drv_flags & WPA_DRIVER_FLAGS_DFS_OFFLOAD)) {
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -703,6 +703,9 @@ struct hostapd_iface {
 	u16 eml_cap;
 	u16 mld_cap;
 #endif
+	struct hostapd_multi_hw_info *multi_hw_info;
+	u8 num_multi_hws;
+	struct hostapd_multi_hw_info *current_hw_info;
 };
 
 /* hostapd.c */
--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -79,6 +79,8 @@ int hostapd_get_hw_features(struct hosta
 	u16 num_modes, flags;
 	struct hostapd_hw_modes *modes;
 	u8 dfs_domain;
+	struct hostapd_multi_hw_info *multi_hw_info;
+	u8 num_multi_hws;
 
 	if (hostapd_drv_none(hapd))
 		return -1;
@@ -141,6 +143,35 @@ int hostapd_get_hw_features(struct hosta
 		}
 	}
 
+	multi_hw_info = hostapd_get_multi_hw_info(hapd, &num_multi_hws);
+	if (multi_hw_info == NULL)
+		return 0;
+
+	hostapd_free_multi_hw_info(iface->multi_hw_info, iface->num_multi_hws);
+	iface->multi_hw_info = multi_hw_info;
+	iface->num_multi_hws = num_multi_hws;
+
+	wpa_printf(MSG_DEBUG, "Multi Underlying Hardwares Info: ");
+
+	for (i = 0; i < num_multi_hws; i++) {
+		struct hostapd_multi_hw_info *hw_info = &multi_hw_info[i];
+		char str[1000];
+		char *pos = str;
+		char *end = pos + sizeof(str);
+		int res;
+
+		for (j = 0; j < hw_info->num_freqs; j++) {
+			res = os_snprintf(pos, end - pos, "%d, ", hw_info->freqs[j]);
+			if (os_snprintf_error(end - pos, res))
+				break;
+			pos += res;
+		}
+
+		*pos = '\0';
+		wpa_printf(MSG_DEBUG, "    %d. hw_idx=%u, frequencies: %s",
+			   i + 1, hw_info->hw_idx, str);
+	}
+
 	return 0;
 }
 
@@ -1324,3 +1355,49 @@ int hostapd_hw_skip_mode(struct hostapd_
 	}
 	return 0;
 }
+
+void hostapd_free_multi_hw_info(struct hostapd_multi_hw_info *multi_hw_info,
+				size_t num_multi_hws)
+{
+	size_t i;
+
+	if (multi_hw_info == NULL)
+		return;
+
+	for (i = 0; i < num_multi_hws; i++)
+		os_free(multi_hw_info[i].freqs);
+
+	os_free(multi_hw_info);
+}
+
+int hostapd_set_current_hw_info(struct hostapd_iface *iface, u32 oper_freq)
+{
+	struct hostapd_multi_hw_info *hw_info;
+	int i, j;
+	bool found = false;
+
+	if (!iface->num_multi_hws)
+		return 0;
+
+	for (i = 0; i < iface->num_multi_hws; i++) {
+		hw_info = &iface->multi_hw_info[i];
+
+		for (j = 0; j < hw_info->num_freqs; j++) {
+			if (hw_info->freqs[j] == oper_freq) {
+				found = true;
+				break;
+			}
+		}
+
+		if (found)
+			break;
+	}
+
+	if (i >= iface->num_multi_hws)
+		return 1;
+
+	iface->current_hw_info = hw_info;
+	wpa_printf(MSG_DEBUG, "Mode: Selected underlying HW: hw_idx=%u",
+		   iface->current_hw_info->hw_idx);
+	return 0;
+}
--- a/src/ap/hw_features.h
+++ b/src/ap/hw_features.h
@@ -28,6 +28,9 @@ int hostapd_prepare_rates(struct hostapd
 void hostapd_stop_setup_timers(struct hostapd_iface *iface);
 int hostapd_hw_skip_mode(struct hostapd_iface *iface,
 			 struct hostapd_hw_modes *mode);
+void hostapd_free_multi_hw_info(struct hostapd_multi_hw_info *multi_macs,
+				size_t num_multi_hws);
+int hostapd_set_current_hw_info(struct hostapd_iface *iface, u32 oper_freq);
 #else /* NEED_AP_MLME */
 static inline void
 hostapd_free_hw_features(struct hostapd_hw_modes *hw_features,
@@ -91,6 +94,17 @@ static inline int hostapd_check_he_6ghz_
 	return 0;
 }
 
+static inline
+void hostapd_free_multi_hw_info(struct hostapd_multi_hw_info *multi_macs,
+				size_t num_multi_hws)
+{
+}
+
+static inline int hostapd_set_current_hw_info(struct hostapd_iface *iface,
+					      u32 oper_freq)
+{
+	return 0;
+}
 #endif /* NEED_AP_MLME */
 
 #endif /* HW_FEATURES_H */
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -329,6 +329,25 @@ struct hostapd_hw_modes {
 	s8 psd_values[NL80211_REG_NUM_POWER_MODES];
 };
 
+/**
+ * struct hostapd_multi_hw_info: Supported multiple underyling hardware info
+ */
+struct hostapd_multi_hw_info {
+	/**
+	 * hw_idx - Hardware index
+	 */
+	u8 hw_idx;
+
+	/**
+	 * num_freqs - Number of entries in the freqs array
+	 */
+	u8 num_freqs;
+
+	/**
+	 * freqs - Array of supported frequencies
+	 */
+	u32 *freqs;
+};
 
 #define IEEE80211_CAP_ESS	0x0001
 #define IEEE80211_CAP_IBSS	0x0002
@@ -5111,6 +5130,18 @@ struct wpa_driver_ops {
 			      const u8 *match, size_t match_len,
 			      bool multicast);
 #endif /* CONFIG_TESTING_OPTIONS */
+
+	/**
+	 * get_multi_hw_info - Get multi hardware data (hardware IDx and
+	 *			supported frequencies)
+	 * @priv: Private driver interface data
+	 * @num_multi_hws: Variable for returning the number of returned hardware
+	 *		    info data
+	 * Returns: Pointer to allocated mutli hardware data on success or
+	 * %NULL on failure. Caller is responsible for freeing this.
+	 */
+	struct hostapd_multi_hw_info * (*get_multi_hw_info)(void *priv,
+							    u8 *num_multi_hws);
 };
 
 /**
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -13747,4 +13747,5 @@ const struct wpa_driver_ops wpa_driver_n
 	.radio_disable = testing_nl80211_radio_disable,
 #endif /* CONFIG_TESTING_OPTIONS */
 	.set_6gclient_type = nl80211_set_6gclient_type,
+	.get_multi_hw_info = nl80211_get_multi_hw_info,
 };
--- a/src/drivers/driver_nl80211.h
+++ b/src/drivers/driver_nl80211.h
@@ -343,5 +343,7 @@ int wpa_driver_nl80211_abort_scan(void *
 int wpa_driver_nl80211_vendor_scan(struct i802_bss *bss,
 				   struct wpa_driver_scan_params *params);
 int nl80211_set_default_scan_ies(void *priv, const u8 *ies, size_t ies_len);
+struct hostapd_multi_hw_info *
+nl80211_get_multi_hw_info(void *priv, u8 *num_multi_hws);
 
 #endif /* DRIVER_NL80211_H */
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -1600,6 +1600,12 @@ struct phy_info_arg {
 	u8 pwr_mode;
 };
 
+struct phy_multi_hw_info_arg {
+	int failed;
+	u8 *num_multi_hws;
+	struct hostapd_multi_hw_info *multi_hws;
+};
+
 static void phy_info_ht_capa(struct hostapd_hw_modes *mode, struct nlattr *capa,
 			     struct nlattr *ampdu_factor,
 			     struct nlattr *ampdu_density,
@@ -2706,4 +2712,142 @@ nl80211_get_hw_feature_data(void *priv,
 	}
 
 	return NULL;
+}
+
+static int phy_multi_hw_info_parse(struct phy_multi_hw_info_arg *multi_hw_info,
+				   struct nlattr *nl_hw_macs)
+{
+	struct nlattr *tb_hw[NL80211_MULTI_HW_MAC_ATTR_MAX + 1];
+	struct nlattr *tb_freq[NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_MAX + 1];
+	struct nlattr *nl_freq_list;
+	struct hostapd_multi_hw_info *multi_hws;
+	int rem_freq_list;
+	u8 hw_idx, freq_idx = 0, i;
+	u32 freq[256];
+
+	nla_parse(tb_hw, NL80211_MULTI_HW_MAC_ATTR_MAX, nla_data(nl_hw_macs),
+		  nla_len(nl_hw_macs), NULL);
+
+	if (!tb_hw[NL80211_MULTI_HW_MAC_ATTR_IDX] ||
+	    !tb_hw[NL80211_MULTI_HW_MAC_ATTR_CHAN_LIST])
+		return NL_STOP;
+
+	hw_idx = nla_get_u8(tb_hw[NL80211_MULTI_HW_MAC_ATTR_IDX]);
+
+	nla_for_each_nested(nl_freq_list, tb_hw[NL80211_MULTI_HW_MAC_ATTR_CHAN_LIST],
+			    rem_freq_list)
+	{
+		nla_parse(tb_freq, NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_MAX,
+			  nla_data(nl_freq_list), nla_len(nl_freq_list), NULL);
+
+		if (!tb_freq[NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_FREQ])
+			return NL_STOP;
+
+		freq[freq_idx++] =
+			nla_get_u32(tb_freq[NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_FREQ]);
+	}
+
+	multi_hws = os_realloc_array(multi_hw_info->multi_hws,
+				     *multi_hw_info->num_multi_hws + 1,
+				     sizeof(*multi_hws));
+	if (!multi_hws) {
+		multi_hw_info->failed = 1;
+		return NL_STOP;
+	}
+
+	multi_hw_info->multi_hws = multi_hws;
+
+	multi_hws = &multi_hw_info->multi_hws[*(multi_hw_info->num_multi_hws)];
+
+	multi_hws->hw_idx = hw_idx;
+	multi_hws->num_freqs = freq_idx;
+
+	multi_hws->freqs = os_calloc(multi_hws->num_freqs, sizeof(u32));
+	if (!multi_hws->freqs) {
+		multi_hw_info->failed = 1;
+		return NL_STOP;
+	}
+
+	for (i = 0; i < multi_hws->num_freqs; i++)
+		multi_hws->freqs[i] = freq[i];
+
+	*(multi_hw_info->num_multi_hws) += 1;
+
+	return NL_OK;
+}
+
+static int phy_multi_hw_info_handler(struct nl_msg *msg, void *arg)
+{
+	struct nlattr *tb_msg[NL80211_ATTR_MAX + 1];
+	struct genlmsghdr *gnlh = nlmsg_data(nlmsg_hdr(msg));
+	struct phy_multi_hw_info_arg *multi_hw_info = arg;
+	struct nlattr *nl_hw_macs;
+	int rem_hw_macs, res;
+
+	nla_parse(tb_msg, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0),
+		  genlmsg_attrlen(gnlh, 0), NULL);
+
+	if (!tb_msg[NL80211_ATTR_MULTI_HW_MACS])
+		return NL_SKIP;
+
+	nla_for_each_nested(nl_hw_macs, tb_msg[NL80211_ATTR_MULTI_HW_MACS],
+			    rem_hw_macs)
+	{
+		res = phy_multi_hw_info_parse(multi_hw_info, nl_hw_macs);
+		if (res != NL_OK)
+			return res;
+	}
+
+	return NL_SKIP;
+}
+
+struct hostapd_multi_hw_info *
+nl80211_get_multi_hw_info(void *priv, u8 *num_multi_hws)
+{
+	u32 feat;
+	struct i802_bss *bss = priv;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	int nl_flags = 0;
+	struct nl_msg *msg;
+	struct phy_multi_hw_info_arg result = {
+		.failed = 0,
+		.num_multi_hws = num_multi_hws,
+		.multi_hws = NULL,
+	};
+
+	*num_multi_hws = 0;
+
+	if (!drv->has_capability)
+		return NULL;
+
+	if (!(drv->capa.flags2 & WPA_DRIVER_FLAGS2_MLO))
+		return NULL;
+
+	feat = get_nl80211_protocol_features(drv);
+	if (feat & NL80211_PROTOCOL_FEATURE_SPLIT_WIPHY_DUMP)
+		nl_flags = NLM_F_DUMP;
+	if (!(msg = nl80211_cmd_msg(bss, nl_flags, NL80211_CMD_GET_WIPHY)) ||
+	    nla_put_flag(msg, NL80211_ATTR_SPLIT_WIPHY_DUMP)) {
+		nlmsg_free(msg);
+		return NULL;
+	}
+
+	if (send_and_recv_msgs(drv, msg, phy_multi_hw_info_handler, &result,
+			       NULL, NULL) == 0) {
+		if (result.failed) {
+			int i;
+
+			for (i = 0; result.num_multi_hws && i < *num_multi_hws; i++)
+				os_free(result.multi_hws[i].freqs);
+
+			os_free(result.multi_hws);
+			*num_multi_hws = 0;
+
+			return NULL;
+		}
+
+		return result.multi_hws;
+	}
+
+	return NULL;
 }
--- a/src/drivers/nl80211_copy.h
+++ b/src/drivers/nl80211_copy.h
@@ -7844,4 +7844,45 @@ enum nl80211_ru_punct_supp_bw {
 	NL80211_RU_PUNCT_SUPP_BW_320,
 };
 
+/**
+ * nl80211_multi_hw_mac_attrs - multi-hw mac attributes
+ *
+ * @NL80211_MULTI_HW_MAC_ATTR_INVALID: invalid
+ * @NL80211_MULTI_HW_MAC_ATTR_IDX: (u8) array index in wiphy @hw_chans to refer an
+ *     underlying hw mac for which the supported channel list is advertised.
+ * @NL80211_MULTI_HW_MAC_ATTR_CHAN_LIST: nested attribute specifying list of
+ *     supported channels, see &enum nl80211_multi_hw_mac_chan_list_attrs
+ * @__NL80211_MULTI_HW_MAC_ATTR_LAST: internal use
+ * @NL80211_MULTI_HW_MAC_ATTR_MAX: maximum multi-hw mac attribute
+ */
+enum nl80211_multi_hw_mac_attrs {
+	__NL80211_MULTI_HW_MAC_ATTR_INVALID,
+
+	NL80211_MULTI_HW_MAC_ATTR_IDX,
+	NL80211_MULTI_HW_MAC_ATTR_CHAN_LIST,
+
+	/* keep last */
+	__NL80211_MULTI_HW_MAC_ATTR_LAST,
+	NL80211_MULTI_HW_MAC_ATTR_MAX =
+				__NL80211_MULTI_HW_MAC_ATTR_LAST - 1,
+};
+
+/**
+ * nl80211_multi_hw_mac_chan_list_attrs - channel attributes for multi-hw
+ *
+ * @__NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_INVALID: invalid
+ * @NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_FREQ: channel center frequency in MHz
+ * @__NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_LAST: internal use
+ * @NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_MAX: maximum channel attribute
+ */
+enum nl80211_multi_hw_mac_chan_list_attrs {
+	__NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_INVALID,
+
+	NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_FREQ,
+
+	/* keep last */
+	__NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_LAST,
+	NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_MAX =
+				__NL80211_MULTI_HW_MAC_CHAN_LIST_ATTR_LAST - 1,
+};
 #endif /* __LINUX_NL80211_H */
--- a/tests/fuzzing/ap-mgmt/ap-mgmt.c
+++ b/tests/fuzzing/ap-mgmt/ap-mgmt.c
@@ -156,6 +156,8 @@ int LLVMFuzzerTestOneInput(const uint8_t
 	hostapd_free_stas(&ctx.hapd);
 	hostapd_free_hw_features(ctx.hapd.iface->hw_features,
 				 ctx.hapd.iface->num_hw_features);
+	hostapd_free_multi_hw_info(ctx.hapd.iface->multi_hw_info,
+				   ctx.hapd.iface->num_multi_hws);
 
 fail:
 	hostapd_config_free(ctx.hapd.iconf);
